balancing/balance_breaker.go

```diff
@@ -0,0 +1,615 @@
+			continue
+		}
+
+		if nodeWeight(node) < nodeWeight(elected) {
+			elected = node
+		}
+	}
+	if elected == nil {
+		return nil, ErrNoActiveNodes
+	}
+	// Disrupt node stats. If all nodes has zero weight only first would
+	// get all the load unless response will come
+	elected.UpdateTimeSpent(time.Since(start))
+	return elected, nil
+}
+
+func inSkipNodes(skipNodes []Node, node Node) bool {
+	for _, skipNode := range skipNodes {
+		if node == skipNode {
+			return true
+		}
+	}
+	return false
+}
+
+// Node is interface of call node
+type Node interface {
+	Calls() float64
+	TimeSpent() float64
+	IsActive() bool
+	SetActive(bool)
+	UpdateTimeSpent(time.Duration)
+}
+
+func nodeWeight(node Node) float64 {
+	return node.TimeSpent()
+}
+
+var (
+	// ErrNoActiveNodes is issued if all nodes are inactive
+	ErrNoActiveNodes = fmt.Errorf("Balancer has no nodes to call")
+)
+
+func newCallMeter(retention, resolution time.Duration) *CallMeter {
+	return &CallMeter{
+		retention:  retention,
+		resolution: resolution,
+		histogram:  newTimeHistogram(retention, resolution, time.Now),
+		now:        time.Now,
+	}
+}
+
+func newCallMeterWithTimer(retention, resolution time.Duration, now func() time.Time) *CallMeter {
+	cm := newCallMeter(retention, resolution)
+	cm.now = now
+	cm.histogram = newTimeHistogram(retention, resolution, now)
+	cm.histogram.now = now
+	return cm
+}
+
+// CallMeter implements Node interface
+type CallMeter struct {
+	retention     time.Duration
+	resolution    time.Duration
+	now           func() time.Time
+	histogram     *histogram
+	inActiveSince time.Time
+	statsShiftMx  sync.Mutex
+}
+
+// UpdateTimeSpent aggregates data about call duration
+func (meter *CallMeter) UpdateTimeSpent(duration time.Duration) {
+	series := meter.histogram.pickSeries(meter.now())
+	if series == nil {
+		return
+	}
+	series.Add(float64(duration), meter.now())
+}
+
+// Calls returns number of calls in last bucket
+func (meter *CallMeter) Calls() float64 {
+	return meter.CallsInLastPeriod(meter.resolution)
+}
+
+// CallsInLastPeriod returns number of calls in last duration
+func (meter *CallMeter) CallsInLastPeriod(period time.Duration) float64 {
+	lastPeriodSeries := meter.histogram.pickLastSeries(period)
+	sum := float64(0)
+	now := meter.now()
+	for _, series := range lastPeriodSeries {
+
+		values := series.ValueRange(now.Add(-period), now)
+		sum += float64(len(values))
+	}
+	return sum
+}
+
+// IsActive aseses if node should be active
+func (meter *CallMeter) IsActive() bool {
+	return meter.inActiveSince == time.Time{}
+}
+
+// SetActive sets meter state
+func (meter *CallMeter) SetActive(active bool) {
+	if meter.IsActive() && !active {
+		meter.inActiveSince = meter.now()
+	}
+	if !meter.IsActive() && active {
+		meter.histogram.shiftData(meter.now().Sub(meter.inActiveSince))
+		meter.inActiveSince = time.Time{}
+	}
+}
+
+// TimeSpent returns float64 repesentation of time spent in execution
+func (meter *CallMeter) TimeSpent() float64 {
```

Comment:
Won't this be too expensive to calculate It happens every time we want to know the node's weight, which is basically every call. If there are a lot rps, then each call we have to traverse each series in a histogram for all of the active nodes to elect a node. 
We could simply aggregate this instead of calculating.